protocol.md

```diff
@@ -726,6 +727,140 @@ Upload-Length: 11
 Upload-Concat: final;/files/a /files/b
 ```
 
+### Challenge
+
+With this extension, Clients can authenticate follow-up requests to a [Creation](#creation) or [Creation With Upload](#creation-with-upload) 
+`POST` request using a high-entropy cryptographic random shared secret - and challenges
+based thereon.
+
+If the Server supports this extension, it MUST add `challenge` to the `Tus-Extension` header.
+
+#### Headers
+
+##### Upload-Secret
+
+The `Upload-Secret` MUST be a high-entropy cryptographic random string, consisting
+of 48 to 256 characters from the [base64 alphabet](https://tools.ietf.org/html/rfc4648#section-4).
+That's 36 to 192 bytes encoded as base64.
+
+The `Upload-Secret` header MUST NOT be included with anything but `POST` requests
+that create a new upload resource. Servers receiving an `Upload-Secret` header with
+any other request MUST respond with a `400 Bad Request` status.
+
+##### Upload-Challenge
+
+If an `Upload-Secret` header was sent with the `POST` request that created
+the upload resource, Clients MUST include an `Upload-Challenge` header with every
+subsequent request targeting that upload resource.
+
+The value of `Upload-Challenge` is defined as follows:
+
+```
+Upload-Challenge = SHA256([Upload-Secret] + SHA256(`[HTTP Method]` + [Upload-Offset] + [Upload-Secret] + [Content-Length]))
```

Comment:
The `Content-Length` header is not always known in advance, for example, when using using `Transfer-Encoding: chunked` with a streaming file to be uploaded. In addition, in some application the handling of these headers are abstracted away and handled by a HTTP library, so the tus client may not know the content length. What should happen in these cases?

Reply:
> The Content-Length header is not always known in advance, for example, when using using Transfer-Encoding: chunked with a streaming file to be uploaded.

A request with `Transfer-Encoding: chunked` should not also contain a `Content-Length` IIRC, so that shouldn't be an issue.

Related: I used `Content-Length` instead of "size of the HTTP request's body" in the proposal because:
- `Content-Length` is only sent with requests where the body length is known in advance, which matters for computing the challenge
- "size of the HTTP request's body" would break any time a request can't be fully sent, as the header with the Challenge would already have been sent and could not anticipate a change in length

> In addition, in some application the handling of these headers are abstracted away and handled by a HTTP library, so the tus client may not know the content length. 

I believe that even where that header is abstracted away by the  HTTP library, the code sending the request would know how many bytes it submitted as the body of that request - and perform computations accordingly.

> What should happen in these cases?

For the `Transfer-Encoding: chunked` case, a value of `0` should be used in place of `Content-Length` for computation.

In cases where a library abstracts away `Content-Length`, but that header is still sent, the client would use the number of bytes it provided as HTTP request body as the value for `Content-Length` in the challenge computations.

Reply:
This is a very valid point from @Acconut. I don't think I have ever in 10 years set the `Content-Length` header manually in .NET. It is added if needed by the underlying http client. Do we need to include `Content-Length` in the algorithm? It seems to me that the only place where we would use something other than a non-zero `Content-Length` is with PATCH requests (and POST for creation with upload but then we won't include a challenge anyway). HEAD and DELETE requests would already provide a zero here. For PATCH requests, the Upload-Offset would change between each request. Would that be enough?

Reply:
I agree with @smatsson that it would be easier if we could not include Content-Length here. I don't think we lose a lot of security here and we could avoid issues in the future with proxies which add or remove Content-Length headers.

Reply:
Another idea, if we want additional security, is to add a kind of salt to the request, e.g. `Upload-Challenge-Salt: <xyz>` which would then be included in the `Upload-Challenge` checksum. 

```
Upload-Challenge-Salt: 1234
Upload-Challenge = SHA256([Upload-Secret] + SHA256(`[HTTP Method]` + [Upload-Offset] + [Upload-Secret] + 1234))
```

What do you think?

Reply:
You're bringing up very good points here. Thanks!

I had assumed that, when a client sends data in the body, it'd also know how much data it is going to send and therefore what would end up in the `Content-Length` header. And when it wouldn't know, it'd use HTTP chunking and not include a `Content-Length` header.

However, I didn't consider the case where the HTTP layer a client builds on might f.ex. always use chunking - or an intermediate layer capturing a HTTP request and - knowing its definitive size - rewrite headers.

The element contributing most to protection is `Upload-Offset` as it ensures that an `Upload-Challenge` can't simply be reused to append arbitrary data - as the offset will have moved on.

However, if we just dropped `Content-Length`, the `Upload-Challenge` used for f.ex. a `HEAD` request (where `Upload-Offset` would be `0`) could be reused for the initial `PATCH` request by an attacker.

This can be mitigated by requiring `none` to be used in place of `Upload-Offset` when calculating the `Upload-Challenge` for a request with no `Upload-Offset` header. Then the `Upload-Challenge`s for a `HEAD` and a `PATCH` at offset `0` would be different.

I originally considered using a salt like `Upload-Challenge-Salt`, too, but had to come to the conclusion that it could just be reused alongside a captured `Upload-Challenge`. So I don't think it'd improve security.

Reply:
Correcting myself on the last sentence: what a `Upload-Challenge-Salt` could definitely achieve is protection against precomputed tables of possible values, so would certainly be a good thing.

Reply:
Replay/reuse of `Upload-Challenge` is something I did not think of. Could we use something like `Challenge-Request-Time` instead of a `Upload-Challenge-Salt` as salt? This way we would get a unique hash each time (and would not need Content-Length etc) while giving the server the possibility to prevent leaked tokens to be used (or at least shorten the time frame they could be used).

Reply:
Sorry my reply took this long!

I don't think `Challenge-Request-Time` would be practical, since HTTP requests on mobile platforms may be queued and sometimes only be sent with great delay.

I also considered a `Challenge-Seed` header, which would be used just like a salt during computation, but whose value simply increments by one with every request. The server could easily keep track of the latest seed it received then, and reject requests with a seed value lower or equal to what it had already received. That would however require coordination between all nodes serving requests, so I'd consider that idea to not be practical either.

So I think a `Upload-Challenge-Salt` would be the best way to go here.

Reply:
What protection would an `Upload-Challenge-Salt` provide? AFAIK, salts prevent the use of dictionary attack (i.e. a list of precomputed hashes of common plain text phrases). However, dictionary attacks are not useful for attacks in the Challenge extension because the `Upload-Secret` already acts as the salt. The secret makes the content of the `Upload-Challenge` header unique for each upload.

Frankly, I don't think we need any additional security mechanisms here. The Challenge extension on its own may not protect against MITM or eavesdropping but I don't think we should be handling this in tus since HTTPS/TLS already takes care of that in a much better way than we could.

Reply:
@felix-schwarz Fair point, did not know that mobile apps does that. :( 

@Acconut Also fair point :D I think the idea was to prevent stealing of existing Upload-Challenge headers and prevent replays, but you are correct in that HTTPS would solve that anyway. Do we want to force the use of TLS when using this extension? I mean you could use it over HTTP as well given that the actual secret is never submitted across the wire (except for when creating the resource).

Reply:
Also to be honest, I would rather see something like `Upload-Salt` where each request contains the header, than using Upload-Offset/Content-Length _sometimes_ and zero sometimes.

Reply:
> Do we want to force the use of TLS when using this extension?

I don't think forcing HTTPS is a good option. A strong recommendation to use it is better IMO

> I would rather see something like Upload-Salt where each request contains the header

Frankly, I still don't understand the benefit of Upload-Salt. If an attack manages to steal the Upload-Challenge header, they likely also have access to the Upload-Salt header. In that case they can reuse the Upload-Challenge and Upload-Salt headers for similar requests. Or there something stopping the reuse of the Upload-Salt that I am missing?

Reply:
The point of using a unique thing (e.g. Upload-Offset/Content-Length/Upload-Salt) is to create a unique hash for each request so that if a Upload-Challenge is stolen it can't be used to re-send requests. Let's say that the algorithm to use did not include anything random like so: `Upload-Challenge = SHA256([Upload-Secret] + [HTTP Method] + [Upload-Secret])`. If I got hold of this Upload-Challenge I could send any request I wish for (this would be equivalent of just sending the Upload-Secret). HTTP Method protects somewhat but not entirely. The idea with adding something random (such as Upload-Salt, Upload-Offset or whatever) is that the hash is then only valid for that particular request. If I get hold of a Upload-Challenge containing the Upload-Salt/Upload-Offset/whatever I could only do minimal of damage as the request is no longer valid. 

As an example:
`Upload-Offset: 1234`
`Upload-Challenge = SHA256([Upload-Secret] + [HTTP Method] + [Upload-Secret] + 1234)`

Sending this request once, when the file's offset is 1234, will be OK. A replay would pass the Upload-Challenge but then fail since the upload offset is (probably) no longer 1234. 

You do have a point that `Upload-Salt` by itself does not protect entirely either as that could also be stolen. That's why I suggested using a timestamp instead, which had issues in mobile apps. For Upload-Salt to work properly the server would have to verify that the salt had not been used before etc. 

We actually use something similar at the bank I work for where we use the following algorithm:
```
sha256({request_body} + {path} + {shared_access_key})
```

In our case we deal with small JSON bodies so that works, but in tus we need something else.

Reply:
Thanks for the detailed explanation, Stefan! As long as the salts would also be reusable, I don't see a benefit in requiring salts. Sadly, we can't reuse the tactic from your bank.

Personally, I see the Challenge extension similar to the `Cookie` header: The cookie mechanism itself does not protect against it being stolen but instead delegates this to the HTTPS protocol which can achieve this goal a lot more effectively.

Let me know if you feel strongly about including additional mechanisms.